[Draft] Blog: Rethinking Software Variants (#1362)
betterscientificsoftware/bssw.io: Articles/Blog/2022-09-SoftwareVariants.md

# Rethinking Software Variants

 **Hero Image:**

  - <img src='../../images/Blog_2209_SoftwareVariants_WarpX.png' />[WarpX: longitudinal electric field in a laser-plasma accelerator, rendered with Ascent and VTK-m.]

#### Contributed by: [Axel Huebl](https://github.com/ax3l)

> #### Publication date: September 27, 2022

Author: Let's update to final release date, probably:
```suggestion
#### Publication date: September 29, 2022
```

### Introduction

Scientific software is often complex, providing many options to control features, supported hardware, and performance.
In this blog post, we will take a closer look at configuration options and their influence on developer and user productivity.

Configuration options of a software package are selected in the installation phase of software: to be precise, the configuration and compilation phase.
As a result, they cannot be changed without reinstalling the software.

> These days, nearly all high-performance computing (HPC) software uses configuration options; and most of the time, these are implemented via preprocessor logic, that is, using #ifdef's in C/C++/Fortran code.

Author: ```suggestion
These days, nearly all high-performance computing (HPC) software uses configuration options; and most of the time, these are implemented via preprocessor logic, that is, using `#ifdef's` in C/C++/Fortran code.
```

Our experience is that problems for developers of "downstream" HPC software arise mainly from the implementation of these options.

For instance, runtime options that change with the chosen binary option need to be carefully documented for users or downstream developers — and they complicate the user experience in already tricky installations.
Workflows have to be established when switching functionality downstream: Do you change configuration of the upstream dependency and rebuild or reinstall? Do you find and use the new variant of the dependency that is packaged separately?
Testing also gets more complicated. If you cannot create a single environment that enables and tests all functionality, you might need to recompile significant portions of the software stack to enable different tests.
Even with sufficient automation, this process increases continuous integration time and can strain deployment resources.

A few specific examples will help illustrate the challenges.

**Breaking public APIs**

> Without due care,  scattered #ifdef switches can easily end up changing public APIs that downstream applications may depend on: extra parameters in functions, different class signatures and constructors, varying members, and so on.

Author: ```suggestion
Without due care, scattered `#ifdef` switches can easily end up changing public APIs that downstream applications may depend on: extra parameters in functions, different class signatures and constructors, varying members, and so on.
```

In the most common case in HPC, the change of the upstream option has to be mirrored 1:1 downstream with `#ifdef`s at call locations, because existing signatures are *changed*.

Such API changes also always lead to incompatible application binary interfaces (ABIs) when building library interfaces.
In the best case, the linker will catch an incompatible binary variant of the same software version via missing symbol errors.

In the worst case, a more subtle ABI break will only manifest at runtime.

> For example, the addition, removal, or change of member variables of public classes via #ifdef's can create undefined behavior when copying or accessing the class, even if these members are private.

Author: ```suggestion
For example, the addition, removal, or change of member variables of public classes via `#ifdef`'s can create undefined behavior when copying or accessing the class, even if these members are private.
```

The website [ABI Laboratory](https://abi-laboratory.pro) summarizes more details on this topic.
For examples that track potentially breaking ABI changes over time, see, for instance, [MPICH](https://abi-laboratory.pro/index.php?view=timeline&l=mpich), [Open MPI](https://abi-laboratory.pro/index.php?view=timeline&l=openmpi) and [c-blosc](https://abi-laboratory.pro/index.php?view=timeline&l=c-blosc).

**The transitive MPI include**

> Adding transitive #include's to third-party software in public APIs is one of the most common mistakes in HPC binary variant design.

Author: ```suggestion
Adding transitive `#include`'s to third-party software in public APIs is one of the most common mistakes in HPC binary variant design.
```

So-called fat binary artifacts can address this problem in part, by compiling multiple backends into the same executable and delegating the code path to choose at runtime.
Unfortunately, few established conventions and tooling currently exist for such an approach, especially across different vendors, so this particular case can be hard to deal with.

### Possible solutions and development policies

The solution to these challenges starts with everyone thinking of themselves as "upstream" developers and thinking about how downstream users might reuse their software.
This even includes application developers: someone might come up with a clever way to integrate an application — like a library — into a larger context, for example, for optimizing ensemble use cases or AI/ML workflows.

We propose the following guidelines or development policies when introducing binary variants into software.

1. Observe  strict extension-only: using compilation variants only for adding *additional* functionality  that *enhances* a package 
    - *Rationale*: toggles dependencies on and off for simplified development and deployment

2. Avoid exclusive compilation options: 

>      - Avoid variants that enable functionality at the cost of disabling another, e.g., #ifdef FOUND_MPI ... #else ....

Author: ```suggestion
     - Avoid variants that enable functionality at the cost of disabling another, e.g., `#ifdef FOUND_MPI ... #else ...`.
```

We changed binary options that control additional field-solvers that require FFTs and linear algebra to provide additional runtime options when enabled, and without changing application behavior if compiled and not used.

More complicated are changes in simulation geometry.
Ideally, WarpX developers would like to offer users a single deployment that provides 1D, 2D, 3D and quasi-cylindrical (RZ) geometry at the same time.
In AMReX, this is a compile-time option.

We are addressing this by progressively compiling all geometries "as 3D" and adding additional 1D and 2D calls to parallel kernel primitives in AMReX 3D interfaces.
Until then, we have increased usability by mirroring the compile time option as runtime user input, which allows us to throw clean error messages, and by compiling multiple runtime libraries per geometry for our Python bindings, which moves the task of a one-time dispatch to the Python level after reading the inputs file.

### Summary

Configuration variants are a common pattern in software design and in HPC software in particular.

> They are often expressed via #ifdef's and seen as an efficient way to reuse and extend already written code.

Author: ```suggestion
They are often expressed via `#ifdef`'s and seen as an efficient way to reuse and extend already written code.
```
